drink/src/runtime/minimal.rs

```diff
@@ -1,65 +1,115 @@
 
+use std::time::SystemTime;
+
+use frame_support::{sp_runtime::traits::Header, traits::Hooks};
+
+/// A helper struct for initializing and finalizing blocks.
+pub struct BlockBuilder<T>(std::marker::PhantomData<T>);
+
+impl<
+        T: pallet_balances::Config + pallet_timestamp::Config<Moment = u64> + pallet_contracts::Config,
+    > BlockBuilder<T>
+{
+    /// Initialize a new block at particular height.
+    pub fn initialize_block(
+        height: frame_system::pallet_prelude::BlockNumberFor<T>,
+        parent_hash: <T as frame_system::Config>::Hash,
+    ) {
+        frame_system::Pallet::<T>::reset_events();
+        frame_system::Pallet::<T>::initialize(&height, &parent_hash, &Default::default());
+        pallet_balances::Pallet::<T>::on_initialize(height);
+        pallet_timestamp::Pallet::<T>::set_timestamp(
+            SystemTime::now()
+                .duration_since(SystemTime::UNIX_EPOCH)
+                .expect("Time went backwards")
+                .as_secs(),
+        );
+        pallet_timestamp::Pallet::<T>::on_initialize(height);
+        pallet_contracts::Pallet::<T>::on_initialize(height);
+        frame_system::Pallet::<T>::note_finished_initialize();
+    }
+
+    /// Finalize a block at particular height.
+    pub fn finalize_block(
+        height: frame_system::pallet_prelude::BlockNumberFor<T>,
+    ) -> <T as frame_system::Config>::Hash {
+        pallet_contracts::Pallet::<T>::on_finalize(height);
+        pallet_timestamp::Pallet::<T>::on_finalize(height);
+        pallet_balances::Pallet::<T>::on_finalize(height);
+        frame_system::Pallet::<T>::finalize().hash()
+    }
+}
+
 /// The macro will generate an implementation of `drink::SandboxConfig` for the given runtime type.
 #[macro_export]
 macro_rules! impl_sandbox_config {
-    (
-        $( #[ $attr:meta ] )*
-        $vis:vis struct $name:ident {
-            runtime: $runtime:tt;
-            default_balance: $default_balance:expr;
-            default_actor: $default_actor:expr;
+	// entry point: generate ext
+    ($name:ident, $runtime:tt, $default_balance:expr, $default_actor:expr) => {
+        $crate::paste::paste! {
+            $crate::impl_sandbox_config!([<EXT_ $name:upper>], $name, $runtime, $default_balance, $default_actor);
         }
-    ) => {
-        $( #[ $attr ] )*
-        $vis struct $name;
-        impl_sandbox_config!($name, $runtime, $default_balance, $default_actor);
     };
-    (
-        $name:ident, $runtime:tt, $default_balance:expr, $default_actor:expr
-    ) => {
-        impl $crate::SandboxConfig for $name {
-            type Runtime = $runtime;
 
-            fn initialize_storage(storage: &mut $crate::frame_support::sp_runtime::Storage) -> Result<(), String> {
+    ($ext:ident, $name:ident, $runtime:tt, $default_balance:expr, $default_actor:expr) => {
+        thread_local! {
+            static $ext: ::std::cell::RefCell<$crate::TestExternalities> = ::std::cell::RefCell::new({
```

Comment:
IMO the addition of `thread_local!` and `RefCell` is a change for the worse. Previously the execution was predictable - the execution was linear and Rust borrow checker was making sure that we use the externalities properly. Now we've changed to internal mutability (note all the changes from `&mut` to `&`) and slapped on the `thread_local!` which slows everything down. 

Can you explain to me why MUST do it this way and cannot do it any other way?

Also, the changes here are backwards incompatible. You've changed the API of public interfaces and structs.

Reply:
probably a middle solution is to keep in `Sandbox` a `dyn Externalities` object...

Reply:
Yes I have to agree that I am not a big fan of the thead_local use myself neither. This was the fastest way for me to test that the solution worked. Let me see if I can find another workaround.